### PR TITLE
Make `write_env_usage` atomic

### PR DESCRIPTION
**Pull request by IanButterworth**
Fixes https://github.com/JuliaLang/Pkg.jl/issues/2633

As per @StefanKarpinski's suggestion in https://github.com/JuliaLang/Pkg.jl/issues/2633#issuecomment-865861720

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
How about also filtering it down either when we load it or when we write it back to only keep the most recent usage of each manifest path? Otherwise I think that as this file gets large this will start to be problematically slow. If you're just appending (which isn't atomic, as we've learned), then it makes sense to keep all the old entries, but since we're not doing that any more, let's clean up as we go.

**Comment by IanButterworth** [comment]
I think that's implemented now. It tidied up my usage file nicely, and sorted it

File: src/Types.jl [hunk 1]
```diff
@@ -459,20 +459,60 @@ end
...
+
+    ## Atomically write usage file
+    while true
+        # read existing usage file
+        usage = if isfile(usage_file)
+            content = read(usage_file)
+            Base.TOML.reinit!(p, String(content); filepath=usage_file)
+            Base.TOML.parse(p)
+        else
+            Dict()
```
Comment (KristofferC) [inline src/Types.jl:477]:
```suggestion
            Dict{String, Any}()
```

File: src/Types.jl [hunk 2]
```diff
@@ -459,20 +459,60 @@ end
...
     usage_file = joinpath(logdir(), usage_filepath)
-    open(usage_file, append=true) do io
-        write(io, entry)
+    timestamp = now()
+
+    ## Atomically write usage file
+    while true
+        # read existing usage file
+        usage = if isfile(usage_file)
+            content = read(usage_file)
```
Comment (KristofferC) [inline src/Types.jl:473]: Isn't this block just `TOML.parse(p, usage_file)`?

Comment (KristofferC) [reply src/Types.jl:473]: I also think it is fine to just use the `TOML.parse(usage_file` API here. In the vast majority of cases you will only read this file twice so it isn't very useful to save the parser object.

File: src/Types.jl [hunk 3]
```diff
@@ -463,15 +463,50 @@ function write_env_usage(source_file::AbstractString, usage_filepath::AbstractSt
...
+            TOML.parsefile(usage_file)
+        else
+            Dict{String, Any}()
+        end
+
+        # record new usage
+        usage[source_file] = [Dict("time" => timestamp)]
+
+        # keep only latest usage info
+        for kv in usage
```
Comment (KristofferC) [inline src/Types.jl:482]: `for k in keys(usage)`?

**Review by KristofferC — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Just want to add that this is really nicely done and should make Pkg much more useable on shared file systems.

**Comment by IanButterworth** [comment]
welp

**Comment by StefanKarpinski** [comment]
I'm pretty sure that all that was required to fix this was adding `force=true` on this line:

https://github.com/JuliaLang/julia/blob/c5f348726cebbe55e169d4d62225c2b1e587f497/base/file.jl#L328

**Comment by IanButterworth** [comment]
That'd be nice and simple!

Before that I'll try to add a multi-process test to Pkg master that reproduces the problem

**Comment by KristofferC** [comment]
> I'm pretty sure that all that was required to fix this was adding force=true on this line:

Nah. See https://github.com/JuliaLang/julia/pull/42253#issuecomment-920335890.

**Pull request merged**